### PR TITLE
Add more Nav natives

### PR DESCRIPTION
**Pull request by KitRifty**
# Fixes:
 - Fixed `CNavArea.SetCostSoFar` not being registered to natives
 - Fixed `CNavArea.GetSizeX/Y` not defined in includes

# Additions:
 - `TheNavMesh.BuildPath`
 - `CNavArea.Get/SetTotalCost`
 - `CNavArea.Get/SetPathLengthSoFar`
 - `CNavArea.ComputePortal`
 - `CNavArea.ComputeClosestPointInPortal`
 - `CNavArea.GetClosestPointOnArea`
 - `CNavArea.ComputeAdjacentConnectionHeightChange`
 - `CNavArea.IsOverlappingPoint`
 - `CNavArea.IsOverlappingArea`
 - `CNavArea.IsOverlappingExtent`
 - `CNavArea.GetDistanceSquaredToPoint`
 - `CNavArea.IsEntirelyVisible`
 - `CNavArea.IsPartiallyVisible`
 - `CNavArea.IsPotentiallyVisible`
 - `CNavArea.IsCompletelyVisible`

~~One thing I'm not sure of is handling `SMPathCost`. Since this exposes `NavAreaBuildPath` as the `TheNavMesh.BuildPath` native, it requires a generic path cost function, but `NavPathCostFunctor` typeset has a `INextBot` argument. I figured that to ensure backwards compatibility, I kept the typeset the same, and in `BuildPath`'s `SMPathCost` object, I pass `nullptr` as the `INextBot` argument.~~ Added `PathFollowerCostFunctor`.

Tested and working on Windows.

### CODE REVIEW

**Comment by Kenzzer** [comment]
Good job on exposing the rest of the natives.
As for the problem you raise regarding `NavPathCostFunctor`, it comes from CBaseNPC early days, when extension used to be a very small library, only focused on exposing pathfollower to its plugin.

Thankfully `NavPathCostFunctor` is a typeset, and there shouldn't be many projects out there making use of it. It should be safe to rename `NavPathCostFunctor` to `PathFollowerNavPathCostFunctor`, without breaking backwards compatibility, and make a new `NavPathCostFunctor` without `INextBot` param.

**Comment by KitRifty** [comment]
I’ll try that then, thanks. `PathFollowerNavPathCostFunctor` seems like a mouthful, though. Would `PathFollowerCostFunctor` be sufficient instead?

File: extension/natives/navarea.h [hunk 1]
```diff
@@ -92,6 +92,95 @@ NAVAREA_NATIVE(GetCostSoFar)
 	return sp_ftoc(pArea->GetCostSoFar());
 }
 
+NAVAREA_NATIVE(SetTotalCost)
+	pArea->SetTotalCost(sp_ctof(params[2]));
+	return 1;
+}
+
+NAVAREA_NATIVE(GetTotalCost)
+	return pArea->GetTotalCost();
```
Comment (Kenzzer) [inline extension/natives/navarea.h:101]: Missing `sp_ftoc` here.

File: extension/natives/navarea.h [hunk 2]
```diff
@@ -92,6 +92,95 @@ NAVAREA_NATIVE(GetCostSoFar)
...
+}
+
+NAVAREA_NATIVE(SetPathLengthSoFar)
+	pArea->SetPathLengthSoFar(sp_ctof(params[2]));
+	return 1;
+}
+
+NAVAREA_NATIVE(GetPathLengthSoFar)
+	return pArea->GetPathLengthSoFar();
```
Comment (Kenzzer) [inline extension/natives/navarea.h:110]: Missing `sp_ftoc` here.

File: extension/natives/navarea.h [hunk 3]
```diff
@@ -92,6 +92,95 @@ NAVAREA_NATIVE(GetCostSoFar)
...
+	VectorToPawnVector(closePosAddr, &closePos);
+	return 1;
+}
+
+NAVAREA_NATIVE(GetDistanceSquaredToPoint)
+	cell_t *posAddr;
+	pContext->LocalToPhysAddr(params[2], &posAddr);
+	Vector pos;
+	PawnVectorToVector(posAddr, &pos);
+	return pArea->GetDistanceSquaredToPoint(pos);
```
Comment (Kenzzer) [inline extension/natives/navarea.h:171]: Missing `sp_ftoc` here.

File: extension/natives/navarea.h [hunk 4]
```diff
@@ -92,6 +92,95 @@ NAVAREA_NATIVE(GetCostSoFar)
...
+}
+
+NAVAREA_NATIVE(ComputeAdjacentConnectionHeightChange)
+	CNavArea *pOther = (CNavArea *)params[2];
+	if (!pOther) 
+	{
+		return pContext->ThrowNativeError("Invalid nav area %x", params[2]);
+	}
+
+	return pArea->ComputeAdjacentConnectionHeightChange(pOther);
```
Comment (Kenzzer) [inline extension/natives/navarea.h:181]: Missing `sp_ftoc` here.

File: extension/natives/navmesh.h [hunk 5]
```diff
@@ -62,6 +62,119 @@ NAVMESHNATIVE(CollectSurroundingAreas)
...
+class SMPathCost : public IPathCost
+{
+public:
+	SMPathCost(IPluginFunction *pFunc) : m_pFunc(pFunc)
+	{
+	}
+
+	virtual float operator()(CNavArea *area, CNavArea *fromArea, const CNavLadder *ladder, const CFuncElevator *elevator, float length) const override final
+	{
+		if (m_pFunc && m_pFunc->IsRunnable())
```
Comment (Kenzzer) [inline extension/natives/navmesh.h:74]: I'm quite unfamiliar with `IPluginFunction` but is the call to `IsRunnable` necessary?

File: scripting/include/cbasenpc/nav.inc [hunk 6]
```diff
@@ -649,4 +852,58 @@ methodmap TheNavAreasVector
...
+ */
+public float Path_Cost(CNavArea area, CNavArea fromArea, CNavLadder ladder, int elevator, float length)
+{
+	if ( fromArea == NULL_AREA )
+	{
+		// At the start of the path
+		return 0.0;
+	}
+
+	static int iTime = 0;
```
Comment (Kenzzer) [inline scripting/include/cbasenpc/nav.inc:866]: Should probably take the chance to delete this as well.

**Review by Kenzzer — CHANGES_REQUESTED** [review]
Overall great, with tiny mistakes that need correcting.

File: extension/natives/navmesh.h (hunk 5 above)
Comment (KitRifty) [reply extension/natives/navmesh.h:74]: I believe so. It's used a few times in SM in similar cases like here:
https://github.com/alliedmodders/sourcemod/blob/1fbe5e1daaee9ba44164078fe7f59d862786e612/core/logic/smn_database.cpp#L242
https://github.com/alliedmodders/sourcemod/blob/1fbe5e1daaee9ba44164078fe7f59d862786e612/core/logic/smn_timers.cpp#L137

Comment (Kenzzer) [reply extension/natives/navmesh.h:74]: Good enough for me!

**Review by Kenzzer — DISMISSED** [review]

**Review by Kenzzer — APPROVED** [review]

**Pull request merged**